docs/streaming.rst

```diff
@@ -6,6 +6,52 @@ Streaming implementation
 This documents describes the algorithm libtorrent uses to satisfy time critical
 piece requests, i.e. streaming.
 
+considerations
+--------------
+
+Streaming data over bittorrent presents several new issues.
+
+1. Sequential downloading hurts download capacity
+	Bittorrent encourages downloaders to do tit-for-tat data transfers between each
+	other. This only works if there is mutual *interest* (each has data the other
+	wants). But if two peers are streaming the torrent in *strictly* sequential
+	piece order, then the "younger" peer has no data to offer the "older" peer.
+	These peers would depend on the upload capacity of any seeds in the swarm.
+2. Full request queues hurt interactivity
+	Whenever the set of time-critical pieces changes (the stream "seeks" to a
+	new position), the optimal move to minimize delay would be to request the
+	newly-critical data from a fast peer *with an empty request queue*.
+3. Non-full request queues hurt throughput
+	Converse to the above, if we want to maximize throughput, we should keep
+	peer request queues as full as possible. In addition to maximizing transfer
+	with a single peer, it's also good to exercise transfer with *all* peers, to discover
+	the *fastest ones*.
+
+Dealing with these issues is an open question.
+
+critical window
+---------------
+
+A common approach to streaming is to use a *critical window* of data ahead of
+the stream position. There are many ways to define such a window.
+
+A window may be *size-based*, where we consider N bytes ahead of the
+stream position to be critical. This has the advantage of being easy to
+understand and configure.
+
+A window may be *time-based*, where we consider N seconds worth of data
+ahead of the stream position to be critical. This requires some way to define
+the "rate" of the stream. For media files, the rate could be the bit rate of the
+media. Otherwise it could be dynamically measured by how fast the consumer
+consumes data.
+
+If the stream consumer consumes data slower than the torrent's download rate
+(this should usually be the case when streaming media), we can decide what to
+do with the "leftover" bandwidth once the critical window is fulfilled. One idea
+is to use classic rarest-first picking to download the rest of the torrent, as
+this can mitigate the problems of sequential downloading, and keeps request
+queues full.
```

Comment:
what's described as a "one idea is to" here is in fact what libtorrent does. I'm worried that phrasing it like this may suggest to the reader that this is not necessarily what libtorrent does.

Reply:
I suppose this is clarified further down

Reply:
I'm using a first-principles approach to documenting streaming logic and functions, because it seems to be a cutting-edge area. I think documenting our basis of understanding is crucial. In contrast, we don't need to document *why* libtorrent uses rarest-first picking, because there's plenty of information available on the topic.

I believe the lack of clarity has left users confused, and having wrong expectations of libtorrent.

Recent user discussions end up disregarding time-critical logic in favor of `sequential_download` ([[1]](https://github.com/arvidn/libtorrent/issues/4408), [[2]](https://github.com/arvidn/libtorrent/pull/5996)). Since time-critical logic is intended to be a better version of `sequential_download`, I conclude either these users don't understand it, or couldn't figure out how to make `set_piece_deadline` work in a context that lacks well-defined deadlines.

Code examples I can find in the wild which use `set_piece_deadline` ([[1]](https://github.com/animeshkundu/pyflix/blob/master/torrent/strategy.py#L28), [[2]](https://github.com/AlexCovizzi/torrenttv/blob/master/torrenttv/bittorrent/file.py#L75-L93), [[3]](https://github.com/FLYKIS2/pelisalacarta-Cmos/blob/master/core/mct.py#L184-L187)) seem to use it wrong. Judging by that code, I *suspect* there's a prevailing belief that the intent of `set_piece_deadline()` is to download a piece *within the next N milliseconds*. Said another way, I think a lot of users are confused between *deadline* and *timeout*.

I think it would be really helpful to link to a project that uses `set_piece_deadline` in the intended way, where deadlines are well-defined, e.g. in a video player. I couldn't find any though.